### PR TITLE
projects: CSV export for projects

### LINKED ISSUE #456: Export search results in CSV
Possibility to export records result from a specific search in a CSV format.

- [ ] Add a button to export records in CSV.
- [ ] Add a flask view to export the records.
- [ ] Create a serializer to output records in CSV.

### PR DESCRIPTION
**Pull request by sebastiendeleze**
* Configures a response handler to export projects record in CSV format.
* Sets the accept mimetype in the `before_request` hook if a `format` argument is present in request.
* Adds a CSV serializer to process records and format them in CSV row.
* Adds the possibility to override the record resource configuration with a dedicated organisation configuration.
* Adds a new stream response, to process a generator and customizes the attachment filename.
* Adds manual translations to translate CSV header.
* Closes #456.

Co-Authored-by: Sébastien Délèze <sebastien.deleze@rero.ch>

### CODE REVIEW

File: sonar/dedicated/hepvs/projects/resource.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+            'realizationFramework', 'funding_funder_type',
+            'funding_funder_name', 'funding_funder_number',
+            'funding_fundingReceived', 'actorsInvolved', 'benefits',
+            'impactOnFormation', 'impactOnProfessionalEnvironment',
+            'impactOnPublicAction', 'promoteInnovation',
+            'relatedToMandate_mandate', 'relatedToMandate_name',
+            'relatedToMandate_briefDescription', 'educationalDocument',
+            'searchResultsValorised'
+        ]),
+                       filename='projects.csv')
```
Comment (zannkukai) [inline sonar/dedicated/hepvs/projects/resource.py:51]: indentation seems strange

File: sonar/dedicated/hepvs/projects/serializers/csv.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+        def transform_external_partners(external_partner):
+            """Convert external partner object into string.
+
+            :param external_partner: External partner dictionary.
+            :returns: String representation of the partner.
+            """
+            text = external_partner['searcherName']
+            if external_partner.get('institution'):
+                text = text + f' ({external_partner["institution"]})'
```
Comment (zannkukai) [inline sonar/dedicated/hepvs/projects/serializers/csv.py:44]:
```python
text = ' '.join([text, external_partner["institution"]])
```

File: sonar/dedicated/hepvs/projects/serializers/csv.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        def transform_actors_involved(actor):
+            """Convert actor object into string.
+
+            :param actor: Actor dictionary.
+            :returns: String representation of the actor.
+            """
+            text = actor['choice'] if actor['choice'] != 'Other' else actor[
+                'other']
+            if actor.get('count'):
+                text = text + f' ({actor["count"]})'
```
Comment (zannkukai) [inline sonar/dedicated/hepvs/projects/serializers/csv.py:56]: same here

File: sonar/resources/projects/resource.py [hunk 4]
```diff
@@ -17,18 +17,33 @@
...
     item_route = f'{list_route}/<pid_value>'
 
+    response_handlers = {
+        'application/json':
+        RecordResponse(JSONSerializer()),
+        'text/csv':
+        StreamResponse(CSVSerializer(csv_included_fields=[
+            'pid', 'name', 'description', 'startDate', 'endDate'
+        ]),
+                       filename='projects.csv')
```
Comment (zannkukai) [inline sonar/resources/projects/resource.py:44]: indentation stange too

File: sonar/resources/projects/serializers/csv.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+                writer.writerow(data)
+                yield line.read()
+
+        return stream_with_context(generate_csv())
+
+    def format_row(self, data):
+        """Format the data for a single row.
+
+        :param data: Data dictionary.
+        """
```
Comment (zannkukai) [inline sonar/resources/projects/serializers/csv.py:72]: !?!? 
Missing code ?

File: sonar/dedicated/hepvs/projects/serializers/csv.py (hunk 2 above)
Comment (jma) [reply sonar/dedicated/hepvs/projects/serializers/csv.py:44]: > ```python
> text = ' '.join([text, external_partner["institution"]])
> ```

Or:
`text = f'{text} ({external_partner["institution"]})'`

File: sonar/dedicated/hepvs/projects/serializers/csv.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            return text
+
+        if data.get('innerSearcher'):
```
Comment (jma) [inline sonar/dedicated/hepvs/projects/serializers/csv.py:59]: You can use a for loop for the next 3 `if`.

File: sonar/resources/projects/serializers/csv.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+            for result in results['hits']['hits']:
+                data = result['metadata']
+                data['pid'] = result['id']
+
+                self.format_row(data)
+
+                # Write CSV data for row.
+                data = self.process_dict(data)
+                writer.writerow(data)
+                yield line.read()
```
Comment (jma) [inline sonar/resources/projects/serializers/csv.py:64]: nice

**Review by jma — APPROVED** [review]

File: sonar/dedicated/hepvs/projects/resource.py (hunk 1 above)
Comment (sebastiendeleze) [reply sonar/dedicated/hepvs/projects/resource.py:51]: This is the formatting done in my IDE...

File: sonar/resources/projects/serializers/csv.py (hunk 5 above)
Comment (sebastiendeleze) [reply sonar/resources/projects/serializers/csv.py:72]: It's a base class than can be extended. And it's a method to do the formatting for a row. 
I putted `pass` statement, but my IDE said I don't have to do that...

File: sonar/resources/projects/serializers/csv.py (hunk 7 above)
Comment (sebastiendeleze) [reply sonar/resources/projects/serializers/csv.py:64]: thanks to @lauren-d !

**Pull request merged**